massa-execution-exports/src/types.rs

```diff
@@ -3,14 +3,158 @@
+        key: Vec<u8>,
+    },
+    /// gets a datastore value (final) for an address, returns ExecutionQueryResponseItem::Bytes(keys) or an error if the address or key is not found
+    AddressDatastoreValueFinal {
+        addr: Address,
+        key: Vec<u8>,
+    },
+
+    /// gets the execution status (candidate) for an operation, returns ExecutionQueryResponseItem::ExecutionStatus(status)
+    OpExecutionStatusCandidate(OperationId),
+    /// gets the execution status (final) for an operation, returns ExecutionQueryResponseItem::ExecutionStatus(status)
+    OpExecutionStatusFinal(OperationId),
+    /// gets the execution status (candidate) for an denunciation, returns ExecutionQueryResponseItem::ExecutionStatus(status)
+    DenunciationExecutionStatusCandidate(DenunciationIndex),
+    /// gets the execution status (final) for an denunciation, returns ExecutionQueryResponseItem::ExecutionStatus(status)
+    DenunciationExecutionStatusFinal(DenunciationIndex),
+
+    /// gets the roll count (candidate) of an address, returns ExecutionQueryResponseItem::RollCount(rolls) or an error if the address is not found
+    AddressRollsCandidate(Address),
+    /// gets the roll count (final) of an address, returns ExecutionQueryResponseItem::RollCount(rolls) or an error if the address is not found
+    AddressRollsFinal(Address),
+    /// gets the deferred credits (candidate) of an address, returns ExecutionQueryResponseItem::DeferredCredits(deferred_credits) or an error if the address is not found
+    AddressDeferredCreditsCandidate(Address),
+    /// gets the deferred credits (final) of an address, returns ExecutionQueryResponseItem::DeferredCredits(deferred_credits) or an error if the address is not found
+    AddressDeferredCreditsFinal(Address),
+
+    // get all information for a given cycle, returns ExecutionQueryResponseItem::CycleInfos(cycle_infos) or an error if the cycle is not found
+    CycleInfos {
+        /// cycle to query
+        cycle: u64,
+        /// optionally restrict the query to a set of addresses. If None, the info for all addresses will be returned.
+        restrict_to_addresses: Option<PreHashSet<Address>>,
+    },
+
+    /// get filtered events. Returns ExecutionQueryResponseItem::Events
+    Events(EventFilter)
+}
+
+/// Execution state query response item
+ pub enum ExecutionQueryResponseItem {
+    /// boolean value
+    Boolean(bool),
+    /// roll counts value
+    RollCount(u64),
+    /// amount value
+    Amount(Amount),
+    /// bytes value
+    Bytes(Vec<u8>),
+    /// vector of bytes value
+    VecBytes(Vec<Vec<u8>>),
+    /// deferred credits value
+    DeferredCredits(BTreeMap<Slot, Amount>),
+    /// execution status value
+    ExecutionStatus(ExecutionQueryExecutionStatus),
+    /// cycle infos value
+    CycleInfos(ExecutionQueryCycleInfos),
+    /// Events
+    Events(Vec<SCOutputEvent>)
+ }
+
+/// Execution status of an operation or denunciation
+ pub enum ExecutionQueryExecutionStatus {
```

Comment:
This enum is a complex, we could make is simpler. This is the perfect case for an array of enums as response. 

executed, success, failed, expired, executable...etc

Reply:
Yes but there are in fact only those 3 mutually exclusive states:
* executed with success
* executed with failure
* either (not executed) or (executed and expired)    (we can't differentiate between those 2)